### PR TITLE
Descriptors: refactor to vulkan-hpp and architecture changes

### PR DESCRIPTION
**Pull request by MoonDweller29**

### CODE REVIEW

File: src/magma/app/render/GBufferResolve.cpp [hunk 1]
```diff
@@ -113,17 +110,17 @@ vk::UniqueSampler GBufferResolve::createImageSampler() {
...
     _descriptorSetLayout.beginSet(0);
     {
-        _descriptorSetLayout.bindCombinedImageSampler(0, gBuffer.getAlbedo().getView(), _imgSampler.get(), VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL);
-        _descriptorSetLayout.bindCombinedImageSampler(1, gBuffer.getNormals().getView(), _imgSampler.get(), VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL);
-        _descriptorSetLayout.bindCombinedImageSampler(2, gBuffer.getGlobalPos().getView(), _imgSampler.get(), VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL);
-        _descriptorSetLayout.bindCombinedImageSampler(3, shadowMapView, shadowMapSampler, VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL);
+        _descriptorSetLayout.bindCombinedImageSampler(0, gBuffer.getAlbedo().getView(), _imgSampler.get());
+        _descriptorSetLayout.bindCombinedImageSampler(1, gBuffer.getNormals().getView(), _imgSampler.get());
+        _descriptorSetLayout.bindCombinedImageSampler(2, gBuffer.getGlobalPos().getView(), _imgSampler.get());
+        _descriptorSetLayout.bindCombinedImageSampler(3, shadowMapView, shadowMapSampler);
```
[the remark this answers is not in the document]
Comment (MoonDweller29) [reply src/magma/app/render/GBufferResolve.cpp:123]: It's hardcoded in this function now.
This argument was the same everywhere, so it seems to be the only possible value there.

File: src/magma/vk/descriptors/DescriptorPool.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+DescriptorPool::DescriptorPool(DescriptorPool &&otherPool) :
+        _device(otherPool._device),
+        _pool(otherPool._pool),
+        _maxSetCount(otherPool._maxSetCount),
+        _setCount(otherPool._setCount)
+{
+    otherPool._device = vk::Device();
+}
```
[the remark this answers is not in the document]
Comment (MoonDweller29) [reply src/magma/vk/descriptors/DescriptorPool.cpp:40]: yes, for storing in std::vector

File: src/magma/vk/descriptors/DescriptorSetInfo.cpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+vk::DescriptorImageInfo *DescriptorSetInfo::newImageInfo() {
+    _imageInfo.emplace_back();
+    return &_imageInfo.back();
+}
+
+vk::WriteDescriptorSet &DescriptorSetInfo::newDescriptorWriteInfo() {
+    _descriptorWrites.emplace_back();
+    return _descriptorWrites[_descriptorWrites.size() - 1]; //@TODO: change to method back
```
[the remark this answers is not in the document]
Comment (MoonDweller29) [reply src/magma/vk/descriptors/DescriptorSetInfo.cpp:15]: just forgot to replace

File: src/magma/vk/descriptors/DescriptorSetLayout.cpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    _descriptorSets.clear();
+    //iteration over existing pools
+    for (uint32_t i = 0; i < _pools.size() && count > 0; ++i) {
+        if (_pools[i].isFull()) {
+            continue;
+        }
+        auto newSets = _pools[i].allocateSets(_layout, count);
+        count -= newSets.size();
+        _descriptorSets.insert(_descriptorSets.end(), newSets.begin(), newSets.end());
+    }
```
[the remark this answers is not in the document]
Comment (MoonDweller29) [reply src/magma/vk/descriptors/DescriptorSetLayout.cpp:39]: Done

File: include/magma/vk/descriptors/DescriptorSetLayout.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    //////////////////////////////////////////////////////////////////////////////////////////////////////
+    void allocateSets(uint32_t count);
+    void beginSet(uint32_t ind);
+    void bindUniformBuffer(uint32_t binding, vk::Buffer buf, vk::DeviceSize offset, vk::DeviceSize range);
+    void bindCombinedImageSampler(uint32_t binding, vk::ImageView imageView, vk::Sampler sampler);
+    void bindStorageImage(uint32_t binding, vk::ImageView imageView, vk::ImageLayout imageLayout);
+    std::vector<vk::DescriptorSet> recordAndReturnSets();
+    //////////////////////////////////////////////////////////////////////////////////////////////////////
+    //
+    //////////////////////////////////////////////////////////////////////////////////////////////////////
```
[the remark this answers is not in the document]
Comment (MoonDweller29) [reply include/magma/vk/descriptors/DescriptorSetLayout.h]: deleted it

File: include/magma/vk/descriptors/DescriptorPool.h [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+class DescriptorPool {
+public:
+    static uint32_t DEFAULT_SET_COUNT;
+
+    DescriptorPool(
+            vk::Device device,
+            const std::unordered_map<vk::DescriptorType, uint32_t> &poolSizes,
+            uint32_t maxSetCount);
+    DescriptorPool(const DescriptorPool &) = delete;
```
[the remark this answers is not in the document]
Comment (MoonDweller29) [reply include/magma/vk/descriptors/DescriptorPool.h]: Added this macro in vulkan_common.h

**Review by aiosadchy — APPROVED** [review]

**Pull request merged**